File: algorithms/medium/execution_of_all_suffix_instructions_staying_in_a_grid.ex
defmodule Solution do
  @spec execute_instructions(n :: integer, start_pos :: [integer], s :: String.t()) :: [integer]
  def execute_instructions(n, start_pos, s) do
    s = String.graphemes(s)

    0..(Enum.count(s) - 1)
    |> Enum.reduce([], fn i, acc -> [Enum.drop(s, i) | acc] end)
    |> Enum.reverse()
    |> Enum.map(fn s -> aux(n, start_pos, s, 0) end)
  end

  def aux(_n, _pos, [], ans), do: ans

  def aux(n, pos, [s | rest], ans) do
    pos = next_pos(pos, s)

    if is_valid_pos?(pos, n), do: aux(n, pos, rest, ans + 1), else: ans
  end

  def next_pos([x, y], "R"), do: [x, y + 1]
  def next_pos([x, y], "L"), do: [x, y - 1]
  def next_pos([x, y], "U"), do: [x - 1, y]
  def next_pos([x, y], "D"), do: [x + 1, y]

  def is_valid_pos?([x, y], n), do: x >= 0 && x < n && y >= 0 && y < n
end

File: algorithms/medium/execution_of_all_suffix_instructions_staying_in_a_grid.rs
impl Solution {
    pub fn execute_instructions(n: i32, start_pos: Vec<i32>, s: String) -> Vec<i32> {
        let pos = (start_pos[0], start_pos[1]);
        let mut ans = vec![0; s.len()];
        let mut s = s.as_bytes();
        for i in 0..s.len() {
            let mut pos = (start_pos[0], start_pos[1]);
            let mut ct = 0;
            for j in i..s.len() {
                pos = match s[j] {
                    b'R' => (pos.0, pos.1 + 1),
                    b'L' => (pos.0, pos.1 - 1),
                    b'U' => (pos.0 - 1, pos.1),
                    b'D' => (pos.0 + 1, pos.1),
                    _ => unreachable!(),
                };
                if pos.0 >= 0 && pos.0 < n && pos.1 >= 0 && pos.1 < n {
                    ct += 1;
                } else {
                    break;
                }
            }
            ans[i] = ct;
        }
        ans
    }
}
